### ISSUE TITLE
Unexpected arguments should print usage

### ISSUE THREAD

**Issue by ysndr** [pacak/bpaf#154]
Sometimes the parser errors could be clearer, if a t least by adding `usage` to the output

Consider this fake git command

```
#[derive(Bpaf, Clone)]
enum Cmd {
	#[bpaf(command)]
	Log {
		#[bpaf(long)]
		oneline: bool
	},

    #[bpaf(hide)]
    Flag
}

#[derive(Bpaf)]
struct Git {
    #[bpaf(long)]
    option: Option<String>,
    #[bpaf(external)]
    cmd: Cmd
}
```

- running `git log --oneline` ✔️
- running `git --oneline log` prints
  ```
  Expected (COMMAND ... | --flag), pass --help for usage information
  ```
  This is technically correct i guess but:
  A) why does `--flag` show up there, its marked as hidden
  B) if `--oneline` is unexpected, what is expected then?
  
Real `git` seems to solve this by also giving you the usage string, ie

```
$ git --oneline log 
unknown option: --oneline
usage: git [-v | --version] [-h | --help] [-C <path>] [-c <name>=<value>]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p | --paginate | -P | --no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           [--super-prefix=<path>] [--config-env=<name>=<envvar>]
           <command> [<args>]
```

That is quite neat as it also shows in which parser the error occured - while parsing the top level command or a subcommand.

In Clap i think i could parse the parse error and print a usage if necessary.
try_parse only has access to Stdout/Stderr, but neither the erroring sub-command or type of error that occured

**Comment by pacak** [comment]
> A) why does `--flag` show up there, its marked as hidden

Great question, this is a bug, will check on that.

> B) if `--oneline` is unexpected, what is expected then?

It expects a command and you can pass `--help` at the end without changing anything else to get exact list. I'll see if I can do this better - dumping `COMMAND` was from before parser had access to full command metadata.

> In Clap i think i could parse the parse error and print a usage if necessary.

You still can do this. You collect args, pass them to the parser. If it succeeds - all is great, if it fails you can append `--help` to args you collected and run the parser again - it will produce help message. At this point you can print both error and help message and exit.

**Comment by ysndr** [comment]
> You still can do this. You collect args, pass them to the parser. If it succeeds - all is great, if it fails you can append --help to args you collected and run the parser again - it will produce help message. At this point you can print both error and help message and exit.

True, that might be an interim solution

i'm not too happy with it printing all the subcommands, might be a bit wordy

**Comment by pacak** [comment]
> I'm not too happy with it printing all the subcommands, might be a bit wordy

Yea, that's the reasoning for the current behavior. Anyway, once I'm done with the documentation generator (hopefully soon, have something that somewhat works locally) - I'll take a look into providing a better suggestion for the example you gave. bpaf can look into each command to see if there's exact match for a flag and suggest to add a command.

**Comment by ysndr** [comment]
> bpaf can look into each command to see if there's exact match for a flag and suggest to add a command.

that would work too, i tried a bit with `anywhere` but that is not supported to derive and would need to be applied to all flags afaict


> Anyway, once I'm done with the documentation generator (hopefully soon, have something that somewhat works locally) - I'll take a look into providing a better suggestion for the example you gave.

👍🏼

**Comment by ysndr** [comment]
Coming back to this, I thought I should add one bit of context: 

> - running `git --oneline log` prints
>  ```
>  Expected (COMMAND ... | --flag), pass --help for usage information
>  ```
>  This is technically correct i guess but:
>  A) why does `--flag` show up there, its marked as hidden
>  B) if `--oneline` is unexpected, what is expected then?

the Important part is that
C) The error does not contain `--oneline` - hence my low effort fix in #155.

**Comment by pacak** [comment]
> C) The error does not contain --oneline - hence my low effort fix in https://github.com/pacak/bpaf/pull/155.

It can't immediately accept `--flag`, since it's part of a command `log` parser so you need to pass `log` first - hence it asking for `COMMAND`.

**Comment by ysndr** [comment]
sure, that part of the error message is ok, what I'm missing is a hint which flag (here: `--oneline`) was given instead that prevented further parsing

**Comment by pacak** [comment]
Yeah. Need to look inside commands to make a hint for missing command. Once
documentation generation stuff is done - will take a look.

**Comment by pacak** [comment]
I did fix all the issues in #174:
- leaking hidden `--flag` to usage
- noting that `log` accepts `--oneline` and not current level

As for printing usage on every error - I'm not sure it's a good idea. `cargo asm --help` output takes more than a screen worth of text in my terminal configuration. I'm assuming it's going to be the same for any moderately complex CLI API. I'm trying to avoid from taking harder to undo decisions by default unless they work as expected in majority scenarios. User can decide if they want to see full usage by passing `--help` as suggested, app author might decide to handle parse failure by slapping `"--help"` at the end of the argument parser and running it again.

**Comment by ysndr** [comment]
> As for printing usage on every error - I'm not sure it's a good idea.

how about short usage?: 

```
Usage: [[-v]... | -q] [--debug]... COMMAND ...
```

**Comment by ysndr** [comment]
anyway i'm happy if the error says what wasn't expected

**Comment by pacak** [comment]
For `cargo-show-asm` it currently look like like this, and that's with a few items suppressed...

```
Usage: [-p SPEC] [--lib | --test TEST | --bench BENCH | --example EXAMPLE | --bin BIN] [--release | --dev | --profile PROFILE] [--target TRIPLE] -C FLAG... -Z FLAG... [--native | --target-cpu CPU] [--rust] [--simplify] -M ARG... [--intel | --att | --llvm | --llvm-input | --mir | --wasm | --mca-intel | --mca-att] [--everything | <ITEM_INDEX> | <FUNCTION> [<INDEX>]]
```
When it fails due to one of the parsers failing (say you must specify asm type) - it will complain what is expected `expected (COMMAND ... | --flag)` (won't suggest hidden `--flag` now). If there's something extra - it will try to guess what this extra might mean and suggest something about that.

Anyway, if you encounter any more strange looking error messages - please open a ticket, will try to make them better.
